src/item_factory.cpp

```diff
@@ -120,9 +120,32 @@ void Item_factory::finalize() {
             }
         }
         // for ammo not specifying loudness (or an explicit zero) derive value from other properties
-        if( obj.ammo && obj.ammo->loudness < 0 ) {
-            obj.ammo->loudness = std::max( std::max( { obj.ammo->damage, obj.ammo->pierce, obj.ammo->range } ) * 3,
-                                           obj.ammo->recoil / 3 );
+        if( obj.ammo ) {
+            if( obj.ammo->loudness < 0 ) {
+                obj.ammo->loudness = std::max( std::max( { obj.ammo->damage, obj.ammo->pierce, obj.ammo->range } ) * 3,
+                                               obj.ammo->recoil / 3 );
+            }
+
+            const auto &mats = obj.materials;
+            if( std::find( mats.begin(), mats.end(), material_id( "hydrocarbons" ) ) == mats.end() &&
+                std::find( mats.begin(), mats.end(), material_id( "oil" ) ) == mats.end() ) {
+                const auto &ammo_effects = obj.ammo->ammo_effects;
+                obj.ammo->cookoff = ammo_effects.count( "INCENDIARY" ) > 0 ||
+                                    ammo_effects.count( "COOKOFF" ) > 0;
+                static const std::set<std::string> special_cookoff_tags = {{
+                    "NAPALM", "NAPALM_BIG",
+                    "EXPLOSIVE_SMALL", "EXPLOSIVE", "EXPLOSIVE_BIG", "EXPLOSIVE_HUGE",
+                    "TOXICGAS", "TEARGAS", "SMOKE", "SMOKE_BIG",
+                    "FRAG", "FLASHBANG"
+                }};
+                obj.ammo->special_cookoff = std::any_of( ammo_effects.begin(), ammo_effects.end(),
```

Comment:
I don't recall it complaining. It might want to align the indents more agressively, possibly pushing `any_of` to next line.